### PR TITLE
Add Home Assistant Discovery

### PR DESCRIPTION
**Pull request by Tenn0**
## Description:
Adds basic Home Assistant Discovery with command argument switch. Use the argument once to setup, can be omitted if its set or not used.
Currently, the only sensor that gets a state is a step sensor from my mi band 3, since I have nothing else to test with.
I´m planning an allow list setting, but i need to think about how to implement that later.
Feel free to tell me about writing less messier code, I´m not a good dev (yet). Also this is my first contribution.
Also there are a bunch of commits because I code on my normal PC, sync code to Github, pull it on my Raspi, build and test there.

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/theengs/gateway/blob/development/docs/participate/development.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by 1technophile** [comment]
Thanks for proposing this!

Here are my comments:
* could you remove from your PR all the files that are not code related (keep only the 2 .py files)
* you can squash your commits into one by using `git rebase -i <sha of the commit before your first one>` and `git push -f`
* theengs decoder can give you directly the [properties](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/src/decoder.h#L38) that are available for a model_id, so you don't need to have those hardcoded
* also I think that the discovery functions could be in a separate python file for readability of the code
* I would prefer that we not submit the discovery every time a sensor is detected, so you will have to store the devices id that has been already submitted
* I'm going to add also comments in the code

File: .vscode/build.bat [hunk 1]
```diff
@@ -0,0 +1,6 @@
+del /F /Q dist
+pip uninstall -y TheengsGateway
+python setup.py sdist
+cd dist
+pip install TheengsGateway-0.1.4.tar.gz --no-deps
+python -m TheengsGateway
```
Comment (1technophile) [inline .vscode/build.bat]: This file should be removed from the PR

File: .vs/VSWorkspaceState.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "ExpandedNodes": [
+    ""
+  ],
+  "SelectedNode": "\\C:\\Users\\tenn0\\Documents\\Projects\\theengsGateway",
+  "PreviewInSolutionExplorer": false
+}
```
Comment (1technophile) [inline .vs/VSWorkspaceState.json]: This file should be removed from the PR

File: .vs/ProjectSettings.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": "x64-Debug"
+}
```
Comment (1technophile) [inline .vs/ProjectSettings.json]: This file should be removed from the PR

File: .vscode/tasks.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    // for the documentation about the tasks.json format
+    "version": "2.0.0",
+    "tasks": [
+        {
+            "label": "build.bat",
+            "type": "shell",
+            "command": "C://Users/tenn0/Documents/Projects/theengsGateway/.vscode/build.bat"
+        }
+    ]
+}
```
Comment (1technophile) [inline .vscode/tasks.json]: This file should be removed from the PR

File: TheengsGateway/__init__.py [hunk 5]
```diff
@@ -32,9 +32,11 @@
     "pass":"",
     "ble_scan_time":5,
     "ble_time_between_scans":5,
-    "publish_topic": "home/TheengsGateway/BTtoMQTT",
+    "publish_topic": "home/TheengsGatewayBTtoMQTT",
```
Comment (1technophile) [inline TheengsGateway/__init__.py]: the topic structure should stay like this so as to preserve compatibility with the [autodiscovery ](https://github.com/1technophile/OpenMQTTGateway/blob/08d8d108d2d7ec4ced0135badb7f54891b8e6100/main/ZmqttDiscovery.ino#L217) function of OpenMQTTGateway:
`"home/TheengsGateway/BTtoMQTT`

This way you can have a setup with ESP32 and Theengs that will leverage the same entities into the home automation controller.

File: build.sh [hunk 6]
```diff
@@ -0,0 +1,6 @@
+rm -r ./dist
+pip3 uninstall -y TheengsGateway
+python3 setup.py sdist
+cd dist
+pip3 install TheengsGateway-0.1.4.tar.gz --no-deps
+curl --data "compiling durch, bruder" http://192.168.2.64:12101/api/text-to-speech
```
Comment (1technophile) [inline build.sh]: remove from the PR please

File: TheengsGateway/ble_gateway.py [hunk 7]
```diff
@@ -125,25 +174,29 @@ def detection_callback(device, advertisement_data):
 
         if data_json:
            gw.publish(data_json, gw.pub_topic + '/' + device.address.replace(':', ''))
+           if gw.discovery == True:
+             gw.publish_device_info(data_json) ## publish sensor data to home assistant mqtt discovery
```
Comment (1technophile) [inline TheengsGateway/ble_gateway.py]: this should be done only one time per sensor (ID)

File: TheengsGateway/__init__.py [hunk 8]
```diff
@@ -32,9 +32,11 @@
...
     "subscribe_topic": "home/TheengsGateway/commands",
-    "log_level": "WARNING"
+    "log_level": "WARNING",
+    "discovery": False,
+    "discovery_topic": "homeassistant/sensor/TheengsGateway"
```
Comment (1technophile) [inline TheengsGateway/__init__.py]: I would suggest to follow the same structure and format as OpenMQTTGateway, here we could stop at `homeassistant/sensor`
 And below you will get all the sensors topics:
image

here is an example of discovery message:
```json
{"stat_t":"+/+/BTtoMQTT/48XXXXXXXA","name":"JQJCY01YM-batt","uniq_id":"48XXXXXXXA-batt","dev_cla":"battery","val_tpl":"{{ value_json.batt | is_defined }}","unit_of_meas":"%","state_class":"measurement","device":{"identifiers":["48XXXXXXXA"],"connections":[["mac","48XXXXXXXA"]],"manufacturer":"Xiaomi","model":"JQJCY01YM","name":"Formaldehyde detector","via_device":"OMG_GAT"}}
```

File: TheengsGateway/ble_gateway.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
 # python 3.6
 
 import asyncio
+from distutils.command.config import config
```
Comment (1technophile) [inline TheengsGateway/ble_gateway.py]: I'm not sure why we need this?

File: TheengsGateway/ble_gateway.py [hunk 10]
```diff
@@ -33,14 +34,21 @@
 from threading import Thread
 
 logger = logging.getLogger('BLEGateway')
+class homeassistant:
+    def __init__(self) -> None:
+        self.device = {}
+        self.device['name'] = "BLEGateway"
+        self.device['identifiers'] = "BLEGateway"
```
Comment (1technophile) [inline TheengsGateway/ble_gateway.py]: We could use the MAC address of the server here so that the ID is unique

**Comment by 1technophile** [comment]
I don't think I'm comprehensive in my review but let's do things step by step

File: TheengsGateway/__init__.py (hunk 8 above)
Comment (Tenn0) [reply TheengsGateway/__init__.py]: I personally would do it this way, to have no mess with sensors. When they are under the /homeassistant/sensir/BTtoMQTT topic, theres a nice way to tell from which gateway it comes and it reduces the clutter. In that case, OMG has to be adjusted to not break compatibility.

File: TheengsGateway/ble_gateway.py (hunk 9 above)
Comment (Tenn0) [reply TheengsGateway/ble_gateway.py]: Same, I´m not sure why VSCode added it without me noticing it.

File: TheengsGateway/__init__.py (hunk 5 above)
Comment (Tenn0) [reply TheengsGateway/__init__.py]: Sorry, typo ftw.

File: TheengsGateway/ble_gateway.py (hunk 7 above)
Comment (Tenn0) [reply TheengsGateway/ble_gateway.py]: So you suggest adding a record of devices already setuped, check if the device is setup yet, and of not send the infos with to the */config topic and if it is already setup, just send attributes and state?

File: TheengsGateway/ble_gateway.py (hunk 10 above)
Comment (Tenn0) [reply TheengsGateway/ble_gateway.py]: Note, this is for the device under HA, the bluetooth device is represented as sensor of the device "BLEGateway". This sorts it under HA, to have less clutter. The sensors have their id as unique_id, from which HA generates its entity_id.

**Comment by Tenn0** [comment]
>     * also I think that the discovery functions could be in a separate python file for readability of the code
Ok, I´ll add the code related to discovery into an own class in a seperate file. But are the args in __init__.py ok?
Edit: I think, for my thoughts to work, I have to restructure the code to have the main run code in a seperate file, because I want my class discovery to extend the class gateway, no need to copy its functions over. But in the current state, it will do circular imports. Any ideas how to solve this?

> theengs decoder can give you directly the [properties](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/src/decoder.h#L38) that are available for a model_id, so you don't need to have those hardcoded

Can you pls explain how to get these? My C++ skills are on the level newbie.

File: TheengsGateway/ble_gateway.py (hunk 7 above)
Comment (1technophile) [reply TheengsGateway/ble_gateway.py]: I would rephrase it like this:
1) when we receive an id
2) check if this id has been discovered or not
3) if yes do nothing
4) if not discovered push all the discovery information (Theengs Decoder will give you all the attributes from the Model_id)

File: TheengsGateway/__init__.py (hunk 8 above)
Comment (1technophile) [reply TheengsGateway/__init__.py]: There is an attribute in the discovery message that gives the gateway that published the message `"via_device":"OMG_GAT"`. This enables to identify from which gateway the publishing message was done.
Gateways are a passthrough for sensors, this structure preserves the unicity of the sensor topics.

**Comment by 1technophile** [comment]
> Can you pls explain how to get these? My C++ skills are on the level newbie.

Np you can refer to this [example](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/examples/python/ScanAndDecode.py#L35)

File: TheengsGateway/__init__.py (hunk 8 above)
Comment (Tenn0) [reply TheengsGateway/__init__.py]: How about a setting for this? I mean, we could let the user decide

Comment (1technophile) [reply TheengsGateway/__init__.py]: if we let it by default to:
`<discovery_prefix>/<component>/<object_id>/config` I'm ok

File: TheengsGateway/ble_gateway.py (hunk 10 above)
Comment (1technophile) [reply TheengsGateway/ble_gateway.py]: I would prefer that we create 2 separate devices, one for the gateway:
image

And one for the sensor that groups all the entities:
image

or only the one for the sensor as the gateway one would not have entities for the moment.

**Comment by Tenn0** [comment]
BTW, is there a nicer way of debugging instead of this everytime I want to test something:
python3 setup.py sdist
cd dist
pip3 install TheengsGateway-0.1.4.tar.gz
?
I mean, since my PC is powerful enough, checking for general problems is no problem, but doing this on the raspi, since my pc has no bluetooth, is wasting time and is annoying, even tho I´m using an automated script to pull my latest sources, build it and execute it.

**Comment by 1technophile** [comment]
I think you could just locate the installation folder of the python script and modify it there, if you do `pip3 uninstall TheengsGateway` it should point you to the folders

**Comment by 1technophile** [comment]
For info I'm adding a filter into OMG so that we don't publish discovery for `IBEACON` and `MS_CDP` as these devices are changing their MAC address regularly and this is creating many sensors into HASS.

**Comment by koenvervloesem** [comment]
> these devices are changing their MAC address regularly and this is creating many sensors into HASS

Couldn't you add a general filter then that doesn't publish discovery of devices with a private address?

**Comment by Tenn0** [comment]
> I think you could just locate the installation folder of the python script and modify it there, if you do `pip3 uninstall TheengsGateway` it should point you to the folders

Easier: On the Raspi, just cd into the project folder, "sudo pip install -e .". Then you can launch right after git pull with "python3 -m TheengsGateway". This setup is only one time needed.

**Comment by 1technophile** [comment]
>Couldn't you add a general filter then that doesn't publish discovery of devices with a private address?

We are publishing the mac_type with OMG that [seems to enable the identification](https://github.com/h2zero/NimBLE-Arduino/blob/96c60fb0dcfcaff204ada4b7809081029ab3671b/src/NimBLEDevice.cpp#L64) as public or private but unfortunately when checking my broker I see that private can also be affected to sensor (even if the mac doesn't change).  Example with a TempoDisc:

```json
{"id":"F3","mac_type":1,"name":"F35","manufacturerdata":"330117550e1017700107042901110100","rssi":-89,"brand":"BlueMaestro","model":"TempoDisc","model_id":"BM_V23","tempc":26.3,"tempf":79.34,"hum":106.5,"dp":27.3,"volt":2.56}
```
Any thoughts @h2zero ?

**Comment by Tenn0** [comment]
> Np you can refer to this [example](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/examples/python/ScanAndDecode.py#L35)



Im having some issues there with my MiBand3. 
[test.py.txt](https://github.com/theengs/gateway/files/8231507/test.py.txt)
In the detection_callback I call 
```
         if data_json:
             print(data_json)
             dev = json.loads(data_json)
             if dev['model_id'] == "MiBand":
               data = getProperties(dev['model_id'])
               print(data)
               data = json.loads(data)
               data = data['properties']
               print(f"{device.attributes()} device")
               print(f"keys: {dev.keys()}")
               print(f"attributes: {getAttribute(dev['model_id'], 'steps')}")
```
But ```getAttribute({getAttribute(dev['model_id'], 'steps')``` returns always ```None```, even when i checked the decoder´s [header for the miband](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/src/devices/Miband_json.h#L8). Should I open a new issue there or am I doing a mistake?

**Comment by 1technophile** [comment]
> But getAttribute({getAttribute(dev['model_id'], 'steps') returns always None, even when i checked the decoder´s [header for the miband](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/src/devices/Miband_json.h#L8). Should I open a new issue there or am I doing a mistake?

Get attribute will enable to retrieve the device attributes like the brand or the model.

If you want to retrieve the step or other property you can call the [getProperties](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/examples/python/ScanAndDecode.py#L34) and just pass the model_id.

**Comment by Tenn0** [comment]
> If you want to retrieve the step or other property you can call the [getProperties](https://github.com/theengs/decoder/blob/fe8541b08275e50e7fcb8c5355c2682e796cc5ec/examples/python/ScanAndDecode.py#L34) and just pass the model_id.

Yeah, that returns a json object. But when I now call 
```
data = getProperties(dev['model_id'])
               print(data)
               data = json.loads(data)
               print(data.keys())
               print(data.values())
```

I get this result: 
```
{"properties":{"steps":{"unit":"int","name":"step-count"}}}
dict_keys(['properties'])
dict_values([{'steps': {'unit': 'int', 'name': 'step-count'}}])
``` 
So the steps property is itself a nested dict with the members 'unit' and 'name', but theres not the actual value stored.

**Comment by 1technophile** [comment]
> So the steps property is itself a nested dict with the members 'unit' and 'name', but theres not the actual value stored.

Indeed this is returning the properties for the discovery like steps, tempc, hum...

The existing gateway code will take care of values.

**Comment by h2zero** [comment]
> > Couldn't you add a general filter then that doesn't publish discovery of devices with a private address?
> 
> We are publishing the mac_type with OMG that [seems to enable the identification](https://github.com/h2zero/NimBLE-Arduino/blob/96c60fb0dcfcaff204ada4b7809081029ab3671b/src/NimBLEDevice.cpp#L64) as public or private but unfortunately when checking my broker I see that private can also be affected to sensor (even if the mac doesn't change). Example with a TempoDisc:
> 
> ```json
> {"id":"F3","mac_type":1,"name":"F35","manufacturerdata":"330117550e1017700107042901110100","rssi":-89,"brand":"BlueMaestro","model":"TempoDisc","model_id":"BM_V23","tempc":26.3,"tempf":79.34,"hum":106.5,"dp":27.3,"volt":2.56}
> ```
> 
> Any thoughts @h2zero ?

We could do this the same way as OMG and just publish the MAC type. I wouldn't want to filter them all because Nordic devices all use a random static MAC address (type = 1,  i.e. switchbot). Better left to the user to filter from the message.

File: TheengsGateway/__init__.py (hunk 8 above)
Comment (Tenn0) [reply TheengsGateway/__init__.py]: Currently, for discovery I publish to `homeassistant/sensor/[unique_id]/[property]` with subtopics for config, attributes and state. Are you ok with this? Or should I only have *homeassistant/sensor/[unique_id]/[propery}/config` with the state and attributes published to corresponding subtopics from `homeassistant/OMG/BTtoMQTT/[unnique_id]/[property] much like OMG? 
BTW, for the Gateway I´ll work on controls, since I´m gonna add an allow-list, atleast for discovery.

Comment (1technophile) [reply TheengsGateway/__init__.py]: I would say to stick per default to OMG structure which is:
`homeassistant/sensor/<object_id>-<property/config`
Example:
`homeassistant/sensor/882D34677766-lux/config`

I don't think you need to have a `state` subtopic as we are not controlling the BLE devices, [referring to this example](https://www.home-assistant.io/docs/mqtt/discovery/#sensors-with-multiple-values).

> BTW, for the Gateway I´ll work on controls, since I´m gonna add an allow-list, atleast for discovery.

You can take inspiration from [this API](https://docs.openmqttgateway.com/use/ble.html#setting-a-white-or-black-list)
Note that this function should be pushed in another PR.

**Comment by 1technophile** [comment]
@Tenn0 do you have any other questions so as to finish this?

**Comment by stepir** [comment]
Hey

I'm currently planning to use TheengsGateway for publishing TPMS sensors of my Camper/RV to my home HomeAssistant (my Camper is always VPN'd to my home).

 Happy to test this too with my sensors!

**Comment by h2zero** [comment]
This will be very useful when combined with #25, if @Tenn0 isn't able to finish it I volunteer to.

**Comment by 1technophile** [comment]
> This will be very useful when combined with https://github.com/theengs/gateway/pull/25, if @Tenn0 isn't able to finish it I volunteer to.

Go ahead @h2zero
